### PR TITLE
Move iterators inside chunk store using MergeSeriesIterator

### PR DESCRIPTION
**Pull request by aaron7**
Depends on #430 - view changes for this PR here: https://github.com/weaveworks/cortex/pull/438/files/83bb69d02a11ef471aebf9408c6a1f1190a6a9da..c38224c193bafc21fe4e38be72c33117893f2dc6

Part of #416 

- Move iterators inside the chunk store
- Create `MergeSeriesIterator` which is created from multiple iterators and implements SeriesIterator.
- Move iterator code to util as it is used across packages and cyclic imports would exist in some cases if it was kept in querier.

### CODE REVIEW

File: pkg/querier/querier.go [hunk 1]
```diff
@@ -221,12 +204,27 @@ func (qm MergeQuerier) RemoteReadHandler(w http.ResponseWriter, r *http.Request)
...
+			}
+
+			// Convert iterators to matrix
+			matrix := make(model.Matrix, 0, len(iterators))
+			for _, it := range iterators {
+				ss := &model.SampleStream{
+					Metric: it.Metric().Metric,
+					Values: it.RangeValues(in),
+				}
+				matrix = append(matrix, ss)
```
Comment (tomwilkie) [inline pkg/querier/querier.go]: This conversion is a little unfortunate.  I think it needs a bit more thought.

File: pkg/util/iterator.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// ValueAtOrBeforeTime implements the SeriesIterator interface.
+func (msit MergeSeriesIterator) ValueAtOrBeforeTime(ts model.Time) model.SamplePair {
+	var closestSamplePair *model.SamplePair
+	var closestTimeDifference time.Duration
+
+	for _, it := range msit.iterators {
+		samplePair := it.ValueAtOrBeforeTime(ts)
+		timeDifference := absTimeDifference(ts, samplePair.Timestamp)
```
Comment (tomwilkie) [inline pkg/util/iterator.go]: Why is this needed?

File: pkg/util/merger.go [hunk 3]
```diff
@@ -27,3 +27,26 @@ func MergeSamples(a, b []model.SamplePair) []model.SamplePair {
 	}
 	return result
 }
+
+// MergeNSamples merges and dedupes n sets of already sorted sample pairs.
+func MergeNSamples(sampleSets ...[]model.SamplePair) []model.SamplePair {
```
Comment (tomwilkie) [inline pkg/util/merger.go]: MergeNSamplesSets?  And MergeSampleSets above?

File: pkg/util/merger.go [hunk 4]
```diff
@@ -27,3 +27,26 @@ func MergeSamples(a, b []model.SamplePair) []model.SamplePair {
...
+	for _, ss := range sampleSets {
+		c <- ss
+	}
+	for ; l > 1; l-- {
+		left, right := <-c, <-c
+		go func() {
+			c <- MergeSamples(left, right)
+		}()
+	}
+	return <-c
```
Comment (tomwilkie) [inline pkg/util/merger.go]: Parallelism not necessary here I think.

File: pkg/util/iterator.go (hunk 2 above)
Comment (aaron7) [inline pkg/util/iterator.go]: Just realised that the samplePair timestamp will always be less than the timestamp given. I will remove.

**Comment by aaron7** [comment]
#442
